### PR TITLE
SpyMaster + Load State Dict

### PR DESCRIPTION
**Pull request by VIGS25**
This PR adds in:
1. SpyMaster class and make TrainedModel use only SpyMaster and Guesser classes
2. Implementation of `load_state_dict`

### CODE REVIEW

File: codenames/supervised/agents.py [hunk 1]
```diff
@@ -92,3 +106,109 @@ def eval(self):
     def train(self):
         for model in self.models:
             model.train()
+
+
+class SpyMaster:
+    """ SpyMaster implementation."""
+
+    name = ['SpyMaster']
```
Comment (oskopek) [inline codenames/supervised/agents.py]: What is this good for?

File: codenames/supervised/agents.py [hunk 2]
```diff
@@ -92,3 +106,109 @@ def eval(self):
...
+    def to(self, device):
+        self.device = device
+        for model in self.models:
+            model.to(device)
+        return self
+
+    def parameters(self):
+        return util.cat_params(*self.models)
+
+    def state_dict(self):
```
Comment (oskopek) [inline codenames/supervised/agents.py]: Same here as for Guesser, make it hierarchical. Maybe even make self.models into a dict, so we have proper names, and not use class names?

File: codenames/supervised/agents.py [hunk 3]
```diff
@@ -92,3 +106,109 @@ def eval(self):
...
+        state_dictionary = {}
+        for model in self.models:
+            for key, value in model.state_dict().items():
+                key = model.__class__.__name__ + '_' + key
+                if key in state_dictionary:
+                    raise ValueError("State dict of Guesser has duplicate keys.")
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
                    raise ValueError("State dict of SpyMaster has duplicate keys.")
```

Comment (oskopek) [reply codenames/supervised/agents.py]: Or even remove the name from the message altogether both here and above in Guesser.

File: codenames/supervised/agents.py [hunk 4]
```diff
@@ -74,16 +81,23 @@ def state_dict(self):
...
                 if key in state_dictionary:
                     raise ValueError("State dict of Guesser has duplicate keys.")
                 else:
                     state_dictionary[key] = value
         return state_dictionary
 
     def load_state_dict(self, state_dict):
-        # TODO(oskopek): Implement deserialization.
-        pass
+        self._create_models()
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
```

Don't need this.

File: codenames/supervised/agents.py [hunk 5]
```diff
@@ -92,3 +106,109 @@ def eval(self):
...
+            for key, value in model.state_dict().items():
+                key = model.__class__.__name__ + '_' + key
+                if key in state_dictionary:
+                    raise ValueError("State dict of Guesser has duplicate keys.")
+                else:
+                    state_dictionary[key] = value
+        return state_dictionary
+
+    def load_state_dict(self, state_dict):
+        self._create_models()
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
```

Don't need this.

File: codenames/supervised/agents.py [hunk 6]
```diff
@@ -92,3 +106,109 @@ def eval(self):
...
+        for model in self.models:
+            for key, value in model.state_dict().items():
+                key = model.__class__.__name__ + '_' + key
+                if key in state_dictionary:
+                    raise ValueError("State dict of Guesser has duplicate keys.")
+                else:
+                    state_dictionary[key] = value
+        return state_dictionary
+
+    def load_state_dict(self, state_dict):
```
Comment (oskopek) [inline codenames/supervised/agents.py]: Same comments as above. Also, let's unify these two implementations into one pair of functions that we call with self.models?

Comment (oskopek) [reply codenames/supervised/agents.py]: Actually, let's make a parent class, which accepts self.models and self.device and implements, `to`, `eval`, `train`, save&load.

File: codenames/supervised/train.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 import codenames.supervised.agents as agents
 import codenames.supervised.data as data
 import codenames.supervised.embeddings as embeddings
-import codenames.supervised.network as network
+# import codenames.supervised.network as network
```
Comment (oskopek) [inline codenames/supervised/train.py]: Just remove it.

**Comment by oskopek** [comment]
Amazing work, thanks! Left some refactoring comments.

File: codenames/supervised/agents.py [hunk 8]
```diff
@@ -74,16 +81,23 @@ def state_dict(self):
...
             for key, value in model.state_dict().items():
-                key = model.__class__.__name__ + key
+                key = model.__class__.__name__ + '_' + key
                 if key in state_dictionary:
                     raise ValueError("State dict of Guesser has duplicate keys.")
                 else:
                     state_dictionary[key] = value
         return state_dictionary
 
     def load_state_dict(self, state_dict):
```
[the remark this answers is not in the document]
Comment (VIGS25) [reply codenames/supervised/agents.py]: Sure, will do that!

[the remark this answers is not in the document]
Comment (VIGS25) [reply codenames/supervised/agents.py]: Does it serialize though?

File: codenames/supervised/agents.py (hunk 4 above)
Comment (VIGS25) [reply codenames/supervised/agents.py]: The torch documentation had initialization before the loading.
Thats why the `_create_models()`

File: codenames/supervised/agents.py (hunk 5 above)
Comment (VIGS25) [reply codenames/supervised/agents.py]: Resolving this since similar discussion above

File: codenames/supervised/agents.py [hunk 9]
```diff
@@ -1,24 +1,76 @@
...
 import torch
+import torch.nn as nn
 
 import codenames.supervised.data as data
 import codenames.supervised.network as network
 import codenames.supervised.util as util
 
 
-class Guesser:
+class Agent(object):
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
class Agent:
```

File: codenames/supervised/agents.py [hunk 10]
```diff
@@ -1,24 +1,76 @@
...
+        for model_name in self.models:
+            self.models[model_name].train()
+
+    def eval(self):
+        for model_name in self.models:
+            self.models[model_name].eval()
+
+
+class Guesser(Agent):
+    """Guesser implementation"""
```
Comment (oskopek) [inline codenames/supervised/agents.py]: Not really an interesting comment :D

File: codenames/supervised/agents.py [hunk 11]
```diff
@@ -1,24 +1,76 @@
...
         self.board_vocab = board_vocab
         self.board_embed = board_embed
         self.clue_vocab = clue_vocab
         self.clue_embed = clue_embed
-        enc_hidden_size = self.clue_embed.embedding_dim
-        self.encoder = network.EncoderOutputRNN(hidden_size=enc_hidden_size, output_size=dec_hidden_size)
-        self.decoder = network.AttnDecoderRNN(dec_hidden_size, dec_hidden_size + self.clue_embed.embedding_dim, data.BOARD_SIZE)
-        self.models = [self.encoder, self.decoder]
+        self.dec_hidden_size = dec_hidden_size
         self.device = "cpu"
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
```

File: codenames/supervised/agents.py [hunk 12]
```diff
@@ -1,24 +1,76 @@
...
 
 import codenames.supervised.data as data
 import codenames.supervised.network as network
 import codenames.supervised.util as util
 
 
-class Guesser:
+class Agent(object):
+
+    def __init__(self, models, device):
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
    def __init__(self, models, device="cpu"):
```

File: codenames/supervised/agents.py [hunk 13]
```diff
@@ -1,24 +1,76 @@
...
         self.clue_embed = clue_embed
-        enc_hidden_size = self.clue_embed.embedding_dim
-        self.encoder = network.EncoderOutputRNN(hidden_size=enc_hidden_size, output_size=dec_hidden_size)
-        self.decoder = network.AttnDecoderRNN(dec_hidden_size, dec_hidden_size + self.clue_embed.embedding_dim, data.BOARD_SIZE)
-        self.models = [self.encoder, self.decoder]
+        self.dec_hidden_size = dec_hidden_size
         self.device = "cpu"
+        self._create_models()
+
+        super(Guesser, self).__init__(models=self.models, device=self.device)
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
        super(Guesser, self).__init__(models=self.models)
```

File: codenames/supervised/agents.py [hunk 14]
```diff
@@ -61,34 +113,72 @@ def train_step(self, board_input, clueword_str, intended_output, criterion, team
...
-        pass
+        # Encoder-Decoder
+        self.enc_hidden_size = enc_hidden_size
+        self.dec_hidden_size = dec_hidden_size
+        self.batch_size = batch_size
 
-    def eval(self):
-        for model in self.models:
-            model.eval()
+        self.device = "cpu"
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
```

File: codenames/supervised/agents.py [hunk 15]
```diff
@@ -61,34 +113,72 @@ def train_step(self, board_input, clueword_str, intended_output, criterion, team
...
+        self._create_models()
 
-    def train(self):
-        for model in self.models:
-            model.train()
+        super(SpyMaster, self).__init__(models=self.models, device=self.device)
```
Comment (oskopek) [inline codenames/supervised/agents.py]:
```suggestion
        super(SpyMaster, self).__init__(models=self.models)
```

File: codenames/supervised/agents.py (hunk 10 above)
Comment (VIGS25) [reply codenames/supervised/agents.py]: :D

**Comment by VIGS25** [comment]
Added the changes suggested. Going to go ahead and merge in.

**Pull request merged**